File: Decompanion/decompanion/dialogs/model_selection_dialog.py
# model_selection_dialog.py

import threading
import javax.swing as swing
from javax.swing import JFrame, JPanel, JButton, JScrollPane, BoxLayout, JLabel, JComboBox
from java.awt import BorderLayout

class ModelSelectionDialog(JFrame):
    """
    Dialog for selecting the Claude model to use.
    """
    def __init__(self, models, default_index, on_selection_complete):
        super(ModelSelectionDialog, self).__init__("Select Claude Model")
        self.models = models
        self.selected_model = None
        self.default_index = default_index
        self.on_selection_complete = on_selection_complete
        self.init_ui()

    def init_ui(self):
        try:
            panel = JPanel()
            panel.setLayout(BoxLayout(panel, BoxLayout.Y_AXIS))

            self.add_instruction_label(panel)
            self.add_model_combo_box(panel)
            self.add_buttons(panel)

            self.getContentPane().add(JScrollPane(panel), BorderLayout.CENTER)
            self.setSize(300, 150)
            self.setLocationRelativeTo(None)
            self.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE)
            self.setVisible(True)
        except Exception as e:
            print("Error initializing ModelSelectionDialog UI: {}".format(e))

    def add_instruction_label(self, panel):
        try:
            instruction_label = JLabel("Select the Claude model to use:")
            instruction_label.setToolTipText("Choose the desired Claude model")
            panel.add(instruction_label)
        except Exception as e:
            print("Error adding instruction label: {}".format(e))

    def add_model_combo_box(self, panel):
        try:
            self.model_combo_box = JComboBox(self.models)
            self.model_combo_box.setSelectedIndex(self.default_index)
            self.model_combo_box.setToolTipText("Select a Claude model from the dropdown")
            panel.add(self.model_combo_box)
        except Exception as e:
            print("Error adding model combo box: {}".format(e))

    def add_buttons(self, panel):
        try:
            button_panel = JPanel()
            ok_button = JButton("OK")
            ok_button.addActionListener(lambda e: self.ok())
            ok_button.setToolTipText("Confirm model selection")
            cancel_button = JButton("Cancel")
            cancel_button.addActionListener(lambda e: self.cancel())
            cancel_button.setToolTipText("Cancel model selection")
            button_panel.add(ok_button)
            button_panel.add(cancel_button)
            self.getContentPane().add(button_panel, BorderLayout.SOUTH)
        except Exception as e:
            print("Error adding buttons to ModelSelectionDialog: {}".format(e))

    def ok(self):
        try:
            self.selected_model = self.model_combo_box.getSelectedItem()
            self.on_selection_complete(self.selected_model)
            self.dispose()
        except Exception as e:
            print("Error in OK action of ModelSelectionDialog: {}".format(e))

    def cancel(self):
        try:
            self.selected_model = None
            self.on_selection_complete(self.selected_model)
            self.dispose()
        except Exception as e:
            print("Error in Cancel action of ModelSelectionDialog: {}".format(e))

def show_model_select_dialog(models):
    """
    Displays the ModelSelectionDialog and waits for user interaction.

    Args:
        models (list): List of available Claude models.

    Returns:
        str or None: The selected model or None if cancelled.
    """
    selected_model = []
    dialog_complete = threading.Event()

    def on_selection(selected):
        selected_model.append(selected)
        dialog_complete.set()

    def create_dialog():
        ModelSelectionDialog(models, 0, on_selection)

    swing.SwingUtilities.invokeLater(create_dialog)
    dialog_complete.wait()
    return selected_model[0]

File: Decyx/decyx/dialogs/ModelSelectionDialog.java
package decyx.dialogs;

import javax.swing.*;
import java.awt.*;
import java.util.List;
import java.util.concurrent.CountDownLatch;

/**
 * Dialog for selecting the Claude model to use.
 */
public class ModelSelectionDialog extends JFrame {

    private final List<String> models;
    private String selectedModel;
    private final Runnable onComplete;
    private JComboBox<String> modelComboBox;

    public ModelSelectionDialog(List<String> models, int defaultIndex, Runnable onComplete) {
        super("Select Claude Model");
        this.models = models;
        this.onComplete = onComplete;
        initUI(defaultIndex);
    }

    private void initUI(int defaultIndex) {
        try {
            JPanel panel = new JPanel();
            panel.setLayout(new BoxLayout(panel, BoxLayout.Y_AXIS));

            addInstructionLabel(panel);
            addModelComboBox(panel, defaultIndex);
            addButtons(panel);

            getContentPane().add(new JScrollPane(panel), BorderLayout.CENTER);
            setSize(300, 150);
            setLocationRelativeTo(null);
            setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            setVisible(true);
        } catch (Exception e) {
            System.out.println("Error initializing ModelSelectionDialog UI: " + e.getMessage());
        }
    }

    private void addInstructionLabel(JPanel panel) {
        JLabel instructionLabel = new JLabel("Select the Claude model to use:");
        instructionLabel.setToolTipText("Choose the desired Claude model");
        panel.add(instructionLabel);
    }

    private void addModelComboBox(JPanel panel, int defaultIndex) {
        modelComboBox = new JComboBox<>(models.toArray(new String[0]));
        modelComboBox.setSelectedIndex(defaultIndex);
        modelComboBox.setToolTipText("Select a Claude model from the dropdown");
        panel.add(modelComboBox);
    }

    private void addButtons(JPanel panel) {
        JPanel buttonPanel = new JPanel();
        JButton okButton = new JButton("OK");
        okButton.addActionListener(e -> ok());
        okButton.setToolTipText("Confirm model selection");
        JButton cancelButton = new JButton("Cancel");
        cancelButton.addActionListener(e -> cancel());
        cancelButton.setToolTipText("Cancel model selection");
        buttonPanel.add(okButton);
        buttonPanel.add(cancelButton);
        getContentPane().add(buttonPanel, BorderLayout.SOUTH);
    }

    private void ok() {
        selectedModel = (String) modelComboBox.getSelectedItem();
        onComplete.run();
        dispose();
    }

    private void cancel() {
        selectedModel = null;
        onComplete.run();
        dispose();
    }

    public String getSelectedModel() {
        return selectedModel;
    }

    /**
     * Displays the dialog and waits for user interaction.
     *
     * @param models list of available Claude models
     * @return the selected model name or null if cancelled
     */
    public static String showDialog(List<String> models) {
        CountDownLatch latch = new CountDownLatch(1);
        ModelSelectionDialog[] dialogHolder = new ModelSelectionDialog[1];

        SwingUtilities.invokeLater(() -> {
            dialogHolder[0] = new ModelSelectionDialog(models, 0, latch::countDown);
        });

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        return dialogHolder[0] != null ? dialogHolder[0].getSelectedModel() : null;
    }
}
